### DISCUSSION TITLE
Backtest visualisation

### DISCUSSION THREAD

**Discussion by jpmediadev** [nautechsystems/nautilus_trader#245] [General]
What do you use to visualize the results as a GUI ?
https://kernc.github.io/backtesting.py/ - this thing looks very nice

**Comment by cjdsellers** [comment]
I've been using Jupyter notebooks with some custom plots. One option is we eventually add some example notebooks, or we could go the route of integrating something more substantial such as a Backtester.py.

It does look quite good. I'll investigate more closely over the next few days and likely open an issue for further discussions.

**Comment by jpmediadev** [comment]
Some examples with your notebooks will be great!

I have idea to use ClickHouse + Grafana for backtest visualisation  and also like realtime dashboard , 
how best to implement it - use Logger subclass with separate thread or write a special class at the strategy level?

**Comment by Troubladore** [comment]
I'm thinking about a UI as well - any progression in thoughts on this space? Was actually contemplating a standard JS/Angular web front end, but I'd prefer to contribute something everyone could use

**Comment by limx0** [comment]
I'm also interesting in this topic.

**Comment by cjdsellers** [comment]
I think we can work towards some more basic backtest visualizations in the same vein as some other open source Python trading platforms currently out there. Note that `empyrical` is currently being used for backtest performance metrics (also used by `pyfolio`).

A JS/Angular front end is a more substantial undertaking, which I had intended to include in a future licensed add-on to NautilusTrader.

The first step would be building a web server which could expose a REST and WebSocket API, leveraging something like FastAPI. This is something an Angular front end could then 'plug into'. Its currently not my intention to add the machinery for this into the open source repo, however I'm happy to discuss topics relating to the above via email.

**Comment by limx0** [comment]
Another good looking library https://github.com/ranaroussi/quantstats

**Reply by cjdsellers** [reply]
This looks _very_ good actually.

**Reply by thematz1** [reply]
I checked this out  (quantstats) and landed on Cerebro ( https://github.com/mementum/backtrader ) from there. I think they accomplish many nice things with a similar idea as Nautilus Trader, but in pure python. 

Could be interesting to take a look! They have an argument to plot the results with Pyfolio or Quantstats after running the test. I also find it interesting how they allow you to easily input any type of bar data (1m aggregations, 1 hour aggregations, etc) which could help to limit the ram required to run a backtest (at the trade-off of accuracy). 

I thought I would share incase anyone wanted to take a peak and get some ideas for feature implementation here.

**Reply by cjdsellers** [reply]
Thanks for sharing @thematz1. I've been aware of Backtrader for a while and have played with it in the past. It doesn't look like the project is too active these days - however I'm aware of many forks which have improved on it, both public and private.

So we currently use the `empyrical` library for backtest performance stats, which is what `pyfolio` calls into under the hood. Unfortunately that library is also not very active, so its prime time we seek an alternative.

Feel free to continue investigating this theme and post any ideas you have! Possibly an integration with a more active and built out project such as `quantstats` could be the way to go?

**Reply by yohplala** [reply]
> Feel free to continue investigating this theme and post any ideas you have! Possibly an integration with a more active and built out project such as `quantstats` could be the way to go?

Hi Christopher,
Following your invitation, I can highlight that `quantstats` is already used by what I identify as Nautylus trader's 'main competitor', [Jesse-AI](https://github.com/jesse-ai).
Their forked repo: https://github.com/jesse-ai/crypto-quantstats

**Reply by thematz1** [reply]
Im wondering how i could just output the returns to CSV in order to simply use the series in one of these visualizers. I was planning on investigating what the input should be from an actual trading strategy into quantstats. I figured posting it here could save some time if someone who has tried reads it. Any ideas?

**Comment by OnlyC** [comment]
While we don't have the perfect solution yet, can you suggest how should I start to build a FastAPI server to expose backtest 
 data, result, indicator's data ...? 
@cjdsellers 
I've done something like this to qtpylib, but then I found nautilus-trader. Now I need to do it again :)

**Comment by DeevsDeevs** [comment]
Hey, @cjdsellers, awesome project, thank you!
Any updates about this visualisation part? i found nothing in docs about it
Are there any dirty solution (if we don't have stable solution for that)

**Comment by KashyapCKotak** [comment]
Hello @limx0 @cjdsellers 

just came across your awesome project and intending to use it, but I am too addicted to backtesting.py 's visualisation (and had added my own code around it). Were we able to put in any solution to visualize either in the library itself or some working dirty solutions?

Waiting for your reply.

**Comment by biiiipy** [comment]
Yeah, nothing beats charts to help understand what the strategy is actually doing. I thought speed and correctness is the most important things, but when starting actually the real problem is visualizing what is happening.

**Comment by aligheshlaghi97** [comment]
Hi everyone,

Inspired by this discussion, I decided to create and deploy a new visualization tool for backtest results using `TradingView`'s `lightweight-charts`, `TypeScript` and `react`. I chose this combination to ensure high performance, smooth rendering (for large amount of data points) and easy integration.

I’d love to hear your thoughts—any feedback, improvement suggestions, or integration ideas are more than welcome!

GitHub repo: https://github.com/Finance-Insight-Lab/bt-visualizer

**Reply by faysou** [reply]
That very interesting. Do you have an example on how to use it with nautilus ?

From what I see you have provided a format to export to in order to use the tool.

I think when I'll need a similar tool I'll try to get it to work in jupyter, and I'll use your work to see how to use lightweight charts.

Thank you !

**Reply by aligheshlaghi97** [reply]
Thanks for the quick response! I'm not very experienced with Nautilus yet, but I'm definitely interested in learning more. I plan to run a backtest with it in the coming days and will share the results afterward.

In the meantime, if you'd like to try out `bt-visualizer`, you'll need to provide four CSV files: one each for OHLC, trades, equity, and stats.
Moreover, `bt-visualizer` is not still some package/extension right now to be used in jupyter or in any languages. Potentially it would be easy to create a `TypeScript` package but rn I don't have any ideas about integrating this with `Python` or other languages.

**Reply by faysou** [reply]
I think lightweight charts already has an extension for jupyter, and its python api mimicks the javascript api.

Also if you want to link javascript to jupyter widgets and python and do non obvious stuff, anywidget is very interesting, I saw videos about it a few weeks ago. anywidget is used already by recent versions of plotly.

**Reply by aligheshlaghi97** [reply]
Thanks a lot for your ideas and sharing them with me. I'll certainly explore them further.

**Reply by faysou** [reply]
No problem, thanks for what you have shared as well.

**Comment by aligheshlaghi97** [comment]
Hi everyone,
I’ve worked on making my backtest visualization tool usable directly in JupyterLab — you can now easily install it with:

```
pip install bt-visualizer
```

@faysou Thanks again for your suggestions! I ended up using AnyWidget to load the built JavaScript and CSS files inside JupyterLab, based on your advice. Really appreciate your insights and guidance.

**Comment by faysou** [comment]
That's nice, thank you, I was actually thinking I could do something similar at some point, but if you've done it already, it's nice. Either it will fit what I need or it will be a start. I'll try what you have done at some point (but not immediately).

**Comment by adamantmm** [comment]
The ClickHouse + Grafana route makes sense once you want a shared ops dashboard across multiple bots/strategies. For single-strategy backtest review I still mostly use Jupyter + Plotly — candlesticks, fill markers, and spread bands tell me more than aggregate equity curves.

One thing that helped on MM-style strategies: overlay intended quotes vs actual fills. Backtests that assume touch fills look great until you see how often you would have been behind the queue in live data.

bt-visualizer looks useful — will try it next time I need something lighter than a full Grafana stack.
